EDU-11546 - Secrets for FastStore (#1917)
vtexdocs/dev-portal-content: docs/faststore/docs/security/managing-secrets.mdx

  %% 4. WebOps determines which provider to use for build/deployment
  WebOps ->> WebOps: Select configured provider

  %% 5. WebOps propagates secret to selected provider and initiates build
  alt Provider A
    WebOps ->> ProviderA: Transmit secret securely
    ProviderA ->> ProviderA: Consume secret, prepare environment
    WebOps ->> ProviderA: Start build/deployment
    ProviderA ->> ProviderA: Build and host application
    ProviderA -->> WebOps: Send status update
  else Provider B
    WebOps ->> ProviderB: Transmit secret securely
    ProviderB ->> ProviderB: Consume secret, prepare environment
    WebOps ->> ProviderB: Start build/deployment
    ProviderB ->> ProviderB: Build and host application
    ProviderB -->> WebOps: Send status update
  else Provider N
    WebOps ->> ProviderN: Transmit secret securely
    ProviderN ->> ProviderN: Consume secret, prepare environment
    WebOps ->> ProviderN: Start build/deployment
    ProviderN ->> ProviderN: Build and host application
    ProviderN -->> WebOps: Send status update
  end

  %% 6. Feedback to user
  WebOps -->> User: Notify build and secret operation success
```

> 

Reviewer: A imagem do diagrama está pequena e dificulta um pouco para entender o fluxo :( 
Como tem a opção de aumentar a imagem na página, não acho que seja um grande problema, mas para ajudar o rápido entendimento do diagrama, tava pensando que a explicação abaixo poderia estar como passos. O que acha? Por exemplo:

1. User initiates request
The user initiates a request to create, update, or delete a secret through the WebOps interface.

2. Permission validation
    a. When the user initiates a request, WebOps first validates the user's permissions and the request scope. 
    b. The scope defines the context, such as project, environment, namespace, or account, that determines where and how the secret will be stored, propagated, and accessed.

4. Secret storage
    a. Once validated, WebOps forwards the request made by the user (create, update, or delete a secret) to AWS Secrets Manager.
    b. AWS confirms the request ("and store the secret"?). 

5. Provider selection
    a. With the AWS confirmation of the request, WebOps identifies the appropriate deployment provider based on the project’s configuration ("project’s configuration, environment settings, and deployment requirements"?).
    b. WebOps securely transmits the secret information to the selected provider.

6. Secret deployment
    a. The chosen provider consumes the secret and prepares the environment.
    b. WebOps triggers the build and the secret deployment process.
    c. The provider builds and hosts the application and returns a status update to WebOps.

7. WebOps notification to the user
    a. Provider sends status update to WebOps.
    b. In the WebOps interface, the user receives confirmation of the successful secret operation and its deployment.

> When a user initiates a request to create or delete a secret, WebOps first validates the user's permissions and the request scope. The scope defines the context, such as project, environment, namespace, or account, that determines where and how the secret will be stored, propagated, and accessed.

Reviewer: - Precisa de algum [perfil de acesso](https://help.vtex.com/pt/tutorial/roles--7HKK5Uau2H6wxE1rH5oRbc?&utm_source=autocomplete) ou [recurso](https://help.vtex.com/pt/tutorial/roles--7HKK5Uau2H6wxE1rH5oRbc?&utm_source=autocomplete#recursos) específico para gerenciar secrets?
- Como é definido esse scope? Acho que não entendi, é na interface do WebOps quando a pessoa está adicionando um secret?

Reviewer: Aqui tem a possibilidade de algum erro durante essa operação? Por exemplo, WebOps começa a fazer a validação das permissões de usuário e o escopo, mas não consegue seguir com a validação (talvez o usuário não tem permissões suficiente para fazer a operação?), dá um erro e mostra uma mensagem na interface para o usuário?

Author: Validei com o time de FS Platform, e não há nenhuma restrição de acesso. O escopo são as próprias configurações da loja que são enviadas para o AWS Secrets Manager, como nome da conta, projeto da FastStore, essas coisas. Vou tentar explicar melhor isso.

Once validated, WebOps forwards the operation to AWS Secrets Manager, which confirms the action. Next, WebOps identifies the appropriate deployment provider based on the project’s configuration. It then securely transmits the secret information to the selected provider.

The chosen provider consumes the secret and prepares the environment, after which WebOps triggers the build and deployment process. The provider proceeds to build and host the application and returns a status update to WebOps.

> Finally, WebOps notifies the user that the secret operation and the deployment were successful.

Reviewer: Há a possibilidade de não ter sucesso na operação? O usuário é notificado também?

> 

Reviewer: Antes de começar a gerenciar os secrets, a pessoa precisa de algum pré-requisito, como perfil de acesso, alguma configuração no projeto...?

## Instructions

To manage your secrets, access your [FastStore WebOps dashboard](https://developers.vtex.com/docs/guides/faststore/1-onboarding-dashboard) and navigate to the **Settings** tab.

In the [Secrets](https://developers.vtex.com/docs/guides/faststore/1-onboarding-dashboard#secrets) section, you can [create](#creating-secrets), [update](#updating-secrets), or [delete](#deleting-secrets) secrets following the steps below.

> 

Reviewer: Talvez adicionar uma imagem da interface dessa seção de Secrets em Settings.

Reviewer 2: Concordo. A pessoa pode acabar pulando direto para as outras seções e não vendo essa parte.

### Creating secrets

> 1. Input the related values in the `Key` and `Value` fields.

Reviewer: ```suggestion
1. In the `Key` and `Value` fields, enter their respective values.

```

Reviewer: Tem algum padrão para escrever esses valores? Seria possível adicionar algum exemplo fictício aqui?

Reviewer 2: Acho que dá para explorar melhor o que são esses campos.
Por exemplo:

In the `Key` field, enter the name of the secret, which serves as its unique identifier (e.g., `VTEX_API_TOKEN`, `NEXT_SECRET_KEY`). In the `Value` field, enter the corresponding sensitive information you want to store (e.g., the actual token, key, or password).

2. Click `Add`. You’ll see the message: `New Secret added successfully. Changes will take effect in the next successful deployment.`

> 3. Click on `Redeploy`. You’ll see the message: `Creating deployment…`, then `Deployment created successfully`.

Reviewer: Talvez ter uma imagem de apoio entre o passo 2 e 3 ou após o passo 3. Esse botão `Redeploy` só aparece após adicionar os valores nos campos Key e Value?

### Updating secrets

1. Go to the `Current Keys` section.
2. Alongside the secret you need to update, click `⋮`, then click `Edit`.

> 3. Update the secret in the modal that will open.

Reviewer: Talvez adicionar um gif ou imagem para apoiar essas instruções.

4. Click `Update`. You’ll see the message: `Secret updated successfully. Changes will take effect in the next successful deployment.`

> 5. Click on `Redeploy`. You’ll see the message: `Creating deployment…`, then `Deployment created successfully`.

Reviewer: Há a possibilidade de falhar essas operações? Se sim, o que o usuário pode fazer?

### Deleting secrets

1. Go to the `Current Keys` section.
2. Alongside the secret you need to update, click `⋮`, then click `Delete`.
3. Validate the secret you want to remove. 

>    >⚠ This action can’t be undone.

Reviewer: Aqui aparece alguma mensagem / pop-up confirmando se a pessoa realmente quer deletar o secret?

In the previous version, an empty `vtex.env` file needed to be at the project's root to enable secrets management. The key-value pairs were stored in the `secrets.revealed.json` file, then encrypted into `secrets.hidden.json` before being committed and pushed to the repository. In the current version, these files were deprecated, and you manage your secrets directly in the WebOps interface.

>⚠️ For local development, use the `vtex.env` file to define environment-specific secrets, such as API keys and tokens, needed to run and test your FastStore project on your local machine. This file is now exclusively for local setups and should always be added to `.gitignore` to prevent version control from tracking it. Since secrets stored in `vtex.env` are not propagated to cloud providers, any secrets you define locally will not be available in deployed environments via WebOps.

In this guide, learn how to manage secrets in your FastStore project using WebOps.

> 

Reviewer 2: ```suggestion

## Production environment

In production, secrets are stored securely using AWS Secrets Manager and are automatically injected into the deployment process by WebOps. This ensures a secure, centralized, and consistent workflow across all supported cloud providers, without exposing secrets in your project code.
```

> ## Secrets flow

Reviewer 2: Acho que esse conteúdo poderia ser uma doc separada porque aqui explica muito mais o funcionamento por trás da feature. Não é necessariamente relevante para quem só quer criar/modificar/excluir um secret

Reviewer 2: Caso ache melhor separar, acho que daria para ter uma seção tipo

## Production environment

In production, secrets are stored securely using AWS Secrets Manager and are automatically injected into the deployment process by WebOps. This ensures a secure, centralized, and consistent workflow across all supported cloud providers, without exposing secrets in your project code.

<details>
<summary>Deployment feedback</summary>

```mermaid
sequenceDiagram
  actor A1 as User
  participant WebOps as WebOps

  %% 5. Feedback to user
  alt Build successful
    WebOps -->> A1: Status notification
  else Build failed
    Note left of WebOps: No message is sent to the user. Check the Deploys tab for logs.
  end
```

</details>

## Instructions

To manage your secrets, access your [FastStore WebOps dashboard](https://developers.vtex.com/docs/guides/faststore/1-onboarding-dashboard) and navigate to the **Settings** tab.

In the [Secrets](https://developers.vtex.com/docs/guides/faststore/1-onboarding-dashboard#secrets) section, you can [create](#creating-secrets), [update](#updating-secrets), or [delete](#deleting-secrets) secrets following the steps below.

### Creating secrets

1. Input the related values in the `Key` and `Value` fields.
2. Click `Add`. A pop-up with `New Secret added successfully` will open. Then, you'll see the message `Secrets have changed. Changes will take effect in the next successful deployment.` alongside a `Redeploy` button.

> 3. Click `Redeploy`. You’ll see the message `Redeploying with secret changes`, then a pop-up with `Deployment created successfully` will open.

Reviewer 2: ```suggestion
3. Click `Redeploy` to redeploy your website with the updated secret configuration. You’ll see the message `Redeploying with secret changes`, then a pop-up with `Deployment created successfully` will open.

```
